autocodesign/localcodesignasset/profilelookup.go

```diff
@@ -0,0 +1,121 @@
+package localcodesignasset
+
+import (
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/bitrise-io/go-xcode/autocodesign"
+	"github.com/bitrise-io/go-xcode/profileutil"
+)
+
+func findProfile(localProfiles []profileutil.ProvisioningProfileInfoModel, platform autocodesign.Platform, distributionType autocodesign.DistributionType, bundleID string, entitlements autocodesign.Entitlements, minProfileDaysValid int, certSerials []string, deviceIDs []string) *profileutil.ProvisioningProfileInfoModel {
+	for _, profile := range localProfiles {
+		if isProfileMatching(profile, platform, distributionType, bundleID, entitlements, minProfileDaysValid, certSerials, deviceIDs) {
+			return &profile
+		}
+	}
+
+	return nil
+}
+
+func isProfileMatching(profile profileutil.ProvisioningProfileInfoModel, platform autocodesign.Platform, distributionType autocodesign.DistributionType, bundleID string, entitlements autocodesign.Entitlements, minProfileDaysValid int, certSerials []string, deviceIDs []string) bool {
+	if !isActive(profile, minProfileDaysValid) {
+		return false
+	}
+
+	if !hasMatchingDistributionType(profile, distributionType) {
+		return false
+	}
+
+	if !hasMatchingBundleID(profile, bundleID) {
+		return false
+	}
+
+	if !hasMatchingPlatform(profile, platform) {
+		return false
+	}
+
+	if !hasMatchingLocalCertificate(profile, certSerials) {
+		return false
+	}
+
+	if !containsAllAppEntitlements(profile, entitlements) {
+		return false
+	}
+
+	if !provisionsDevices(profile, deviceIDs) {
+		return false
+	}
+
+	return true
+}
+
+func hasMatchingBundleID(profile profileutil.ProvisioningProfileInfoModel, bundleID string) bool {
+	return profile.BundleID == bundleID
+}
+
+func hasMatchingLocalCertificate(profile profileutil.ProvisioningProfileInfoModel, localCertificateSerials []string) bool {
+	var profileCertificateSerials []string
+	for _, certificate := range profile.DeveloperCertificates {
+		profileCertificateSerials = append(profileCertificateSerials, certificate.Serial)
+	}
+
+	return 0 < len(intersection(localCertificateSerials, profileCertificateSerials))
+}
+
+func containsAllAppEntitlements(profile profileutil.ProvisioningProfileInfoModel, appEntitlements autocodesign.Entitlements) bool {
+	profileEntitlements := autocodesign.Entitlements(profile.Entitlements)
+	hasMissingEntitlement := false
+
+	for key, value := range appEntitlements {
+		profileEntitlementValue := profileEntitlements[key]
+
+		// The project entitlement values can have variables coming from build settings which will be resolved later
+		// during the archive action. It is not the best but this is also the logic used at other places. An example of
+		// what we could be comparing:
+		// 		$(AppIdentifierPrefix)${BASE_BUNDLE_ID}.ios == 72SA8V3WYL.io.bitrise.samples.fruta.los
+		if reflect.DeepEqual(profileEntitlementValue, value) == false {
```

Comment:
In the existing code there is a special check for iCloud containers (https://github.com/bitrise-io/go-xcode/commit/981b8b2b6833ae2e0f4f9793beedf30b857e8c88), I assume this passes if there are extra containers in the local profile (vs the app).
In theory we could check the existence of app Entitlements one-by-one and ignore the ones with envs.

Reply:
Reused `autocodesign.FindMissingContainers`.